src/Scripting/Core/ScriptBuilder.cs

```diff
@@ -66,7 +66,7 @@ public int GenerateSubmissionId(out string assemblyName, out string typeName)
             int id = Interlocked.Increment(ref _submissionIdDispenser);
             string idAsString = id.ToString();
             assemblyName = _assemblyNamePrefix + "-" + idAsString;
-            typeName = "Submission#" + idAsString;
+            typeName = "Submission_" + idAsString;
```

Comment:
This change was due to a CLS-compliance check by `XmlSerializer`.
It could also help for scenarios such as [this](https://stackoverflow.com/q/38646619).

Reply:
What CLS compliance check?

Reply:
```
  Message: 
    System.ArgumentException : Identifier 'Submission#0' is not CLS-compliant.
    Parameter name: ident
  Stack Trace: 
    CodeIdentifier.CheckValidIdentifier(String ident)
    CodeIdentifier.EscapeKeywords(String identifier, CodeDomProvider codeProvider, StringBuilder sb)
    CodeIdentifier.GetCSharpName(Type t, Type[] parameters, Int32 index, StringBuilder sb)
    CodeIdentifier.GetCSharpName(Type t, Type[] parameters, Int32 index, StringBuilder sb)
    CodeIdentifier.GetCSharpName(Type t)
    TypeDesc.get_CSharpName()
    SourceInfo.CastTo(TypeDesc td)
    XmlSerializationWriterILGen.WriteElements(SourceInfo source, String enumSource, ElementAccessor[] elements, TextAccessor text, ChoiceIdentifierAccessor choice, String arrayName, Boolean writeAccessors, Boolean isNullable)
    XmlSerializationWriterILGen.WriteMember(SourceInfo source, String choiceSource, ElementAccessor[] elements, TextAccessor text, ChoiceIdentifierAccessor choice, TypeDesc memberTypeDesc, Boolean writeAccessors)
    XmlSerializationWriterILGen.GenerateTypeElement(XmlTypeMapping xmlTypeMapping)
    XmlSerializationWriterILGen.GenerateElement(XmlMapping xmlMapping)
    TempAssembly.GenerateRefEmitAssembly(XmlMapping[] xmlMappings, Type[] types, String defaultNamespace, Evidence evidence)
    TempAssembly.ctor(XmlMapping[] xmlMappings, Type[] types, String defaultNamespace, String location, Evidence evidence)
    XmlSerializer.GenerateTempAssembly(XmlMapping xmlMapping, Type type, String defaultNamespace)
    XmlSerializer.ctor(Type type, String defaultNamespace)
    XmlSerializer.ctor(Type type)
```

Reply:
Also, it's not even a legal C# name, which raises problems referring to script types (as mentioned in the SO link above).

Reply:
That's intentional. These types are implementation details, we don't want code to refer to them.

Reply:
@tmat those types are not visible within a script anyhow, but IMO we shouldn't prevent users from using them in more advanced scenarios. What's the downside?

Reply:
The downside is that we wouldn't be able to change the implementation if we needed to.

Reply:
Even if the types were legal C# names, relying on those names would still be an implementation detail. It's very different from a public API or a compiler-generated symbol.

I think C# scripts should not be limited to what the can run, and this poses unnecessary limitations.